### PR TITLE
[8.0] Pilot stamps available in the pilot execution environment

### LINKED ISSUE #6077: PilotLogger - How to identify pilots in an early submission state
### Discussed in https://github.com/DIRACGrid/DIRAC/discussions/6075

<div type='discussions-op-text'>

<sup>Originally posted by **martynia** May 13, 2022</sup>
Hi,

In my work on remote pilot logging I identified the following problem: 
  In the original extended (==remote) logger, which is now being removed, pilots were identified by a UUID. The UUID was created by ExtendedLogger and saved to a file, so it could be accessed easily. I used the UUID to identify pilot files shipped to Tornado. From an old description by the original author this was supposed to be a reliable way to identify pilots. But even then, the UUID generated this way had nothing to do with the pilot reference stored in the PilotDB. In the current setup pilot reference is known only when ConfigureSite command is run, so quite late. This created a problem, when pilot dies early - we have no way to identify it. On the other hand even in an early stage the pilot reference is displayed as a frame title in the WebApp, so there must be a way to get it early. One of the possibility would be moving the code of `__setFlavour` method from ConfigureSite command to SiteDirector and use it, but it looks like a cumbersome solution. Or create a thin command which is run first and just gets required info from the CE ?

Thanks, JM</div>

### PR DESCRIPTION
**Pull request by atsareg**
This PR is for passing pilot stamps to the pilot execution environment. This should be done differently for different ComputingElements. 

- [x] HTCondorCE
- [x] ARC/ARC6
- [x] AREX
- [x] SSH
- [x] Cloud

BEGINRELEASENOTES

*Resources
NEW: [ALL]ComputingElements - pass pilot stamp to the pilot environment

ENDRELEASENOTES

### CODE REVIEW

**Comment by sfayer** [comment]
Hi,

I'm not sure the cloud changes will work quite correctly: The pilot/inst UUID is just a random value generated by the cloud provider used as the pilotReference, it doesn't connect back to the instRandom which is only currently included in the instance name. (The return statement from the end of submitPilot has also gone missing).

It is possible to get the name from inside the VM, but I think it's probably safer (more portable to different cloud providers) to substitute it directly into the metadata, something like this:
https://github.com/DIRACGrid/DIRAC/commit/77ae2f06eeaed42ae2a7cc421e3aba6712d8a66f

Regards,
Simon

**Comment by atsareg** [comment]
Thanks Simon, I have included your proposed changes

**Comment by martynia** [comment]
Hi,
In the meantime I tested the stamps for HT Condor, and I can confirm that the pilot stamps are correctly passed to the pilot.

File: src/DIRAC/Resources/Computing/BatchSystems/Condor.py [hunk 1]
```diff
@@ -106,15 +107,15 @@ def submitJob(self, **kwargs):
     Output = $(Cluster).$(Process).out
     Error = $(Cluster).$(Process).err
     Log = test.log
-    Environment = CONDOR_JOBID=$(Cluster).$(Process)
+    Environment = "CONDOR_JOBID=$(Cluster).$(Process) DIRAC_PILOT_STAMP=$(stamp)"
```
Comment (fstagni) [inline src/DIRAC/Resources/Computing/BatchSystems/Condor.py:108]:
```suggestion
    Environment = "CONDOR_JOBID=$(Cluster).$(Process) DIRAC_PILOT_STAMP=$(stamps)"
```

File: src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py [hunk 2]
```diff
@@ -197,7 +198,7 @@ def __writeSub(self, executable, nJobs, location, processors):
 output = $(Cluster).$(Process).out
 error = $(Cluster).$(Process).err
 log = $(Cluster).$(Process).log
-environment = "HTCONDOR_JOBID=$(Cluster).$(Process)"
+environment = "HTCONDOR_JOBID=$(Cluster).$(Process) DIRAC_PILOT_STAMP=$(stamp)"
```
Comment (fstagni) [inline src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py:196]: is `stamp` defined?

Comment (andresailer) [reply src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py:196]: `stamp` is defined in the `Queue stamp...` below

File: src/DIRAC/Resources/Computing/ARCComputingElement.py [hunk 3]
```diff
@@ -256,6 +256,7 @@ def _writeXRSL(self, executableFile, inputs=None, outputs=None, executables=None
 (inputFiles=({executable} "{executableFile}") {xrslInputAdditions})
 (stdout="{diracStamp}.out")
 (stderr="{diracStamp}.err")
+(environment="(DIRAC_PILOT_STAMP {diracStamp}"))
```
Comment (sfayer) [inline src/DIRAC/Resources/Computing/ARCComputingElement.py]: Janusz has been testing this and it seems this line is causing submission errors; should it be:
`(environment=("DIRAC_PILOT_STAMP {diracStamp}"))` (e.g. with the bracket outside of the string after the equals)?

Comment (sfayer) [reply src/DIRAC/Resources/Computing/ARCComputingElement.py]: Hmm, my previous suggestion isn't quite right... The key and value should also be quoted independently:
`(environment=("DIRAC_PILOT_STAMP" "{diracStamp}"))`

Comment (martynia) [reply src/DIRAC/Resources/Computing/ARCComputingElement.py]: I can confirm that the submission  to ARC CE works now when setting:
`(environment=("DIRAC_PILOT_STAMP" "{diracStamp}"))`

File: src/DIRAC/Resources/Computing/BatchSystems/Condor.py (hunk 1 above)
Comment (atsareg) [reply src/DIRAC/Resources/Computing/BatchSystems/Condor.py:108]: I do not think this is correct. This is similar to the HTCondorCE case, the stamp is defined in the Queue command below

**Comment by atsareg** [comment]
No extra intervention needed for AREXCE. Please, review

**Review by fstagni — APPROVED** [review]
Review OK for me, but other opinions are more than welcome!

I will merge it before the Thursday hackathon so it can be tried out in DIRAC certification. Some specific test/checks to make?

**Comment by aldbr** [comment]
I just reviewed the PR: it looks OK to me, I have nothing to report.

**Pull request merged**